File: alg/src/main/java/hashTable/singleNum.java
package hashTable;

import java.util.*;

public class singleNum {
    public int solutionWithSet (int [] nums) {
        Set<Integer> set = new HashSet<>();
        int sum = 0;
        int setSum = 0;
        for (int i = 0; i < nums.length; ++i) {
            if (!set.contains(nums[i])) {
                set.add(nums[i]);
                sum += nums[i];
                setSum += nums[i];
            } else {
                sum += nums[i];
            }

        }
        return 2 * setSum - sum;
    }

    public int solutionWithXOR (int [] nums) {
        int result = 0;
        for (int i = 0; i < nums.length; ++i) {
            result = result ^ nums[i];
        }

        return result;
    }
}

File: c++/singleNum.cc
#include <vector>
#include <iostream>

using std::vector;

class Solution {
    public:
        int singleNum (vector<int> & nums) {
            int ret = 0;
            for (auto & i: nums) {
                ret ^= i;
            }
            return ret;
        }
};

int main() {
    vector<int> test{1,2,3,4,4,3,2};
    int result;
    Solution s;
    result = s.singleNum(test);
    std::cout << result << std::endl;
}